src/plugins/shape-generator/index.ts

```diff
@@ -0,0 +1,407 @@
+import { PluginController } from "#plugins/PluginController.js";
+import { ConfigItem } from "..";
+import { ellipseGeneratorExpressions, ellipseLatex } from "./latex/ellipse";
+import {
+  rectangleGeneratorExpressions,
+  rectangleLatex,
+} from "./latex/rectangle";
+import { DispatchedEvent } from "#globals";
+
+interface NewDropdownItem {
+  ariaLabel: string;
+  label: string;
+  okBtnExprId: string;
+  handler: () => void;
+  okBtnHandler: (expr: Desmos.ExpressionState) => void;
+}
+
+export default class ShapeGenerator extends PluginController<{
+  showSliders: boolean;
+}> {
+  static id = "shape-generator" as const;
+  static enabledByDefault = false;
+  static config: readonly ConfigItem[] = [
+    {
+      type: "boolean",
+      key: "showSliders",
+      default: false,
+    },
+  ];
+
+  private _isEditingShape = false;
+
+  private _onDispatchHandler = "";
+
+  private readonly newDropdownItems: NewDropdownItem[] = [
+    {
+      ariaLabel: "Add ellipse",
+      label: "ellipse",
+      okBtnExprId: "shape-generator-ellipse",
+      handler: () => {
+        this.calc.setExpressions(ellipseGeneratorExpressions);
+      },
+      okBtnHandler: (expr) => {
+        const xHelper = this.calc.HelperExpression({
+          latex: "x_{ellipseGenerator}",
+        });
+        const yHelper = this.calc.HelperExpression({
+          latex: "y_{ellipseGenerator}",
+        });
+        const rxHelper = this.calc.HelperExpression({
+          latex: "r_{xEllipseGenerator}",
+        });
+        const ryHelper = this.calc.HelperExpression({
+          latex: "r_{yEllipseGenerator}",
+        });
+        const angleHelper = this.calc.HelperExpression({
+          latex: "A_{ellipseGenerator}",
+        });
+
+        angleHelper.observe("numericValue", () => {
```

Comment:
This seems to assume that `angleHelper` gets its `observe` triggered only after all helpers have their numeric value resolved. While that seems to be correct, a comment to this effect could help future debugging.